File: 2352-equal-row-and-column-pairs/2352-equal-row-and-column-pairs.cpp
class Solution {
public:
    int equalPairs(vector<vector<int>>& grid) {
        map<string,int> rowMap, colMap;  
        int answer = 0; 
        for(int i = 0; i < grid.size(); i++){
            string row = ""; 
            for(int j = 0; j < grid[0].size(); j++){
                row  += to_string(grid[i][j]) + '!'; 
            }
            //cout << row << endl; 
            rowMap[row]++; 
        }

        for(int j = 0; j < grid[0].size(); j++){
            string col = ""; 
            for(int i = 0; i < grid.size(); i++){
                col += to_string(grid[i][j]) + '!'; 
            }
            //cout << col << endl; 
            answer += rowMap[col]; 
        }
        return answer; 
    }
};

File: 2352-equal-row-and-column-pairs/2352-equal-row-and-column-pairs.java
class Solution {
    public int equalPairs(int[][] grid) {
        int n = grid.length; 
        Map<String,Integer> rowMap = new HashMap<>(); 
        Map<String,Integer> colMap = new HashMap<>(); 
        
        for(int i = 0; i < n; i++){
            StringBuilder row = new StringBuilder(); 
            StringBuilder col = new StringBuilder(); 
            for(int j = 0; j < n; j++){
                row.append(String.valueOf(grid[i][j])).append(","); 
                col.append(String.valueOf(grid[j][i])).append(","); 
            }
            rowMap.merge(row.toString(),1,Integer::sum); 
            colMap.merge(col.toString(),1,Integer::sum); 
        }

        int answer = 0; 
        for(var e: rowMap.entrySet()){
            if(colMap.containsKey(e.getKey())) answer += (e.getValue() * colMap.get(e.getKey())); 
        }

        return answer; 
    }
}